### DISCUSSION TITLE
What is the intended way of getting the value of manually registered option

### DISCUSSION THREAD

**Discussion by yngvedh** [ajalt/clikt#514] [Q&A] [answered]
Hi,

I've created a cli app which dynamically sets up its subcommands and options based on a configuration file.
It sets up the commands like so:
```
registerOption(option(parameterName)
                        .prompt(default = value)
                        .help(description))
```
But I cannot figure out a straight forward way to get the value in `run()`. The only way I've found to get the parameters is something like this:
```
registeredOptions().map {
    when(it) {
            is OptionWithValues<*, *, *> -> Pair(it.names.first(), it.value)
            ...
```
I have a few issues with this approach:
* How do I know that option is of type `OptionWithValues<*,*,*>`. it has worked so far, but I fear this can break in a future update. This also is dependent of implementation details of Clikt which I should not need to care about.
* When I fetch the value (`it.value`), it will throw an exception if the parameter has not been supplied. I would love for the value to just be nullable and return null if it is not specified.

Is there a better way of getting the values from the Clikt supplied Options? 

Additionally, I can only think of one other way to make it work, which is to create my own implementation of Option and supply that. The problem with this approach is that it is not easy to implement and seems to require much more knowledge about the inner workings of Clikt than I possess, or than I care to learn as a library user. It would be great if there was some documentation or perhaps some facilities in Clikt, like a base class, making it easier to implement.

Please advice.

**Comment by ajalt** [comment, marked as answer]
- You can use [OptionDelegate][1] instead of `OptionWithValues`, since that's the interface that defines `value`. If it's not an `OptionDelegate`, then it doesn't have a value.
- Options normally do return `null` if they weren't provided. `value` only throws an exception on access when you haven't called `parse`/`main`, or if you access it after `parse` thrown an exception.

```kotlin
class C : CliktCommand() {
    override fun run() {
        registeredOptions().forEach {
            echo("${it.names.joinToString()}: ${(it as? OptionDelegate<*>)?.value}")
        }
    }
}
with(C()) {
    registerOption(option("--foo"))
    main(emptyArray())
}
```

```
$ ./c
--foo: null
-h, --help: false
```

[1]: https://ajalt.github.io/clikt/api/clikt/com.github.ajalt.clikt.parameters.options/-option-delegate/index.html?query=interface%20OptionDelegate%3CT%3E%20:%20GroupableOption,%20ReadOnlyProperty%3CParameterHolder,%20T%3E%20,%20PropertyDelegateProvider%3CParameterHolder,%20ReadOnlyProperty%3CParameterHolder,%20T%3E%3E
